### PR TITLE
reworked owr_init

### PR DESCRIPTION
**Pull request by Rugvip**
see #143

### CODE REVIEW

**Comment by sdroege** [comment]
I don't think this really improves the API, even if it solves your problem. See my specific proposal here: https://github.com/EricssonResearch/openwebrtc/issues/143#issuecomment-77816753

**Comment by Rugvip** [comment]
The idea was that owr_init would be the "do it" function for initializing owr, while owr_init_in_main_context would be used if you want more fine-grained control. I did experiment with a owr_init_in_thread which would init owr and start the mainloop in the current thread, but that didn't seem to add any value.

I guess another, and perhaps better alternative is to simply remove owr_init, and that I instead add some helper functions to the java bindings (which is what this really is about).

File: owr/owr.c [hunk 1]
```diff
@@ -241,59 +237,78 @@ void owr_init()
...
-    owr_main_context_is_external = !!owr_main_context;
-
-    if (!owr_main_context_is_external)
-        owr_main_context = g_main_context_new();
+    owr_main_context = main_context;
 
-    if (owr_main_context_is_external)
-        return;
+    if (!owr_main_context)
+        owr_main_context = g_main_context_default();
```
Comment (sdroege) [inline owr/owr.c]: g_main_context_get_thread_default()

File: owr/owr.c [hunk 2]
```diff
@@ -241,59 +237,78 @@ void owr_init()
...
+void owr_run(void)
 {
-    g_return_if_fail(!owr_initialized);
-    g_return_if_fail(main_context);
+    g_return_if_fail(owr_main_context);
 
-    owr_main_context = main_context;
-    owr_init();
+    owr_main_loop = g_main_loop_new(owr_main_context, FALSE);
+    g_main_loop_run(owr_main_loop);
```
Comment (sdroege) [inline owr/owr.c]: g_main_context_unref() after it ran

Comment (sdroege) [inline owr/owr.c]: g_main_loop_unref() after it ran

File: owr/owr.c [hunk 3]
```diff
@@ -241,59 +237,78 @@ void owr_init()
...
-#endif
-    g_return_val_if_fail(!data, NULL);
-    g_main_context_push_thread_default(owr_main_context);
-    g_main_loop_run(owr_main_loop);
-    g_main_context_pop_thread_default(owr_main_context);
-    return NULL;
+    g_return_if_fail(owr_main_loop);
+
+    g_main_loop_quit(owr_main_loop);
+    g_main_loop_unref(owr_main_loop);
```
Comment (sdroege) [inline owr/owr.c]: And not unref it here

File: owr/owr.c [hunk 4]
```diff
@@ -241,59 +237,78 @@ void owr_init()
...
+    owr_main_loop = NULL;
```
Comment (sdroege) [inline owr/owr.c]: And don't unref/set to NULL here. It's safer to do it all after the main loop is done running for real

**Comment by Rugvip** [comment]
updated with some fixes and a test

**Comment by Rugvip** [comment]
Verified on OS X and Android

**Comment by superdump** [comment]
I think this looks good but I want to port all the example apps over to use it and then land everything in one go.

**Pull request closed without merging**